Change OS config loading behaviour

eskyuu opened the pull request:
Change OS config loading behaviour to:
 - Move OS YAML config loading into the main config PHP code
 - Remove the need to manually load OS YAML files before the config can be accessed
 - Change the method for tracking OS YAML files that have been loaded into config
 - Add a new config merge function that will correctly overwrite lists
 - Remove the need to manually update the OS cache file

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [ ] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [ ] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [ ] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

eskyuu committed:
Change OS config loading behaviour
 - Move OS config loading into the config file
 - Remove the need to manually load OS YAML files before the config can be accessed
 - Add a new method for tracking OS definitions that have been read from YAML
 - Add a new config merge function that will correctly overwrite lists
 - Remove the need to manually update the OS cachs.

eskyuu committed:
StyleCI

eskyuu commented:
@murrant Any ideas why this is failing the tests?

murrant commented on app/ConfigRepository.php:
```diff
@@ -106,6 +113,11 @@ public function get($key, $default = null): mixed
         }
 
+        // Check if we need to load the OS YAML file
+        if (preg_match('/^os\.(?<os>[^.]+)/', $key, $matches)) {
```
Probably could be optimized a little with starts_with() and substr()

eskyuu replied:
Done

murrant commented on app/ConfigRepository.php:
```diff
@@ -293,6 +289,11 @@ public function has($key): bool
         }
 
+        // Check if we need to load the OS YAML file
+        if (preg_match('/^os\.(?<os>[^.]+)/', $key, $matches)) {
```
Probably could be optimized a little with starts_with() and substr()

eskyuu replied:
Done

murrant commented on tests/bootstrap.php:
```diff
@@ -83,6 +83,5 @@
-\LibreNMS\Util\OS::updateCache(true); // Force update of OS Cache
-
 app()->terminate(); // destroy the bootstrap Laravel application
+app()->make('librenms-config');
```
Not sure this is needed, however, I'm unsure if tests clear singleton instances or not.

eskyuu replied:
This was a test to try and figure out why we were getting the test errors.

eskyuu committed:
Remove regex for OS from config code

eskyuu committed:
Use substr() instead of explode() to fetch OS

murrant commented on app/ConfigRepository.php:
```diff
@@ -106,6 +113,12 @@ public function get($key, $default = null): mixed
 
+        // Check if we need to load the OS YAML file
+        if (str_starts_with($key, 'os.')) {
+            $os = substr($key, 3, (strpos($key, '.', 3) ?: strlen($key)) - 3);
```
Dang, that is a bit more convoluted than I thought.  It might be better to do preg_match after doing the str_starts_with.  It might be a little odd, but I do want to be mindful of performance in this code as it is called a lot.

eskyuu replied:
The other option is:
$os = explode('.', $key)[1];

Is preg_match still preferred over explode()?

murrant commented on app/ConfigRepository.php:
```diff
@@ -155,16 +168,9 @@ public function getDeviceSetting($device, $key, $global_prefix = null, $default
-                return $this->config['os'][$os][$key];
-            }
-
             $os_key = "os.$os.$key";
```
This may have a bug in it if $key has a . in it.  That probably shouldn't happen, but it would behave different in that case.

murrant replied:
or maybe the old code had a bug, and this is correct :D

eskyuu replied:
I assume this would mean a key in an OS YAML file with a "." in the key name.  I can't see any examples of these in the yaml files.

murrant commented on app/ConfigRepository.php:
```diff
@@ -185,23 +191,14 @@ public function getCombined(?string $os, string $key, string $global_prefix = ''
 
-        if (! $this->has("os.$os.$key")) {
-            return (array) $this->get($global_key, $default);
+        if ($globalVal || $osVal) {
```
This has a bug if $globalVal is falsy

eskyuu committed:
Fix for getCombined()

murrant committed:
Simplify and use regular cache
I checked and we are not even close to DB cache size limit
However, cache invalidation hooks were removed and likely need to be restored.

murrant committed:
Fix os loading and a couple of other small changes

eskyuu committed:
Use os_def_cache_time config option for OS def cache

eskyuu committed:
Explicit use of DB driver for OS definition cache, and invalidate cache on debug/verbose run

eskyuu committed:
Fix case for function name in test file

eskyuu commented:
@murrant Is this good to merge now?

murrant committed:
Remove redundant os caching

murrant committed:
Fix whitespace

murrant committed:
Fix Config load order
install_dir was not loaded yet

murrant committed:
update comment

murrant committed:
Add missing imports

murrant committed:
Add missing imports
